browser/ui/browser_commands.cc

```diff
@@ -1,29 +1,97 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "brave/browser/ui/browser_commands.h"
 
+#include <utility>
+
 #include "brave/browser/tor/tor_profile_service.h"
 #include "brave/browser/tor/tor_profile_service_factory.h"
 #include "chrome/browser/profiles/profile.h"
 #include "chrome/browser/profiles/profile_manager.h"
 #include "chrome/browser/profiles/profile_window.h"
 #include "chrome/browser/ui/browser.h"
 #include "chrome/browser/ui/tabs/tab_strip_model.h"
+#include "components/web_cache/browser/web_cache_manager.h"
 #include "content/public/browser/browser_thread.h"
 #include "content/public/browser/navigation_controller.h"
+#include "content/public/browser/render_process_host.h"
+#include "content/public/browser/storage_partition.h"
 #include "content/public/browser/web_contents.h"
+#include "services/network/public/cpp/features.h"
 
+using content::BrowserContext;
 using content::BrowserThread;
 using content::NavigationController;
+using content::StoragePartition;
 using content::WebContents;
 
 namespace {
-void NewTorIdentityCallback(WebContents* current_tab) {
+// Reload the page once we have a new circuit and cleared browsing data.
+void NewTorIdentityCallbackReload(WebContents* current_tab) {
   NavigationController& controller = current_tab->GetController();
   controller.Reload(content::ReloadType::BYPASSING_CACHE, true);
 }
+
+// Flush everything else in the storage partition.
+void NewTorIdentityCallbackClearData(WebContents* current_tab) {
+  auto* context = current_tab->GetBrowserContext();
```

Comment:
I'm not sure any of this is necessary because I believe we're already using a new partition/network context when cycling, but this not only has race conditions, it will also cause problems for other tabs because it deletes data across the entire browser context, not just the individual tab